modules/cli/src/main/scala/scala/cli/commands/Package.scala

```diff
@@ -127,23 +127,47 @@ object Package extends ScalaCommand[PackageOptions] {
     logger: Logger,
     outputOpt: Option[String],
     force: Boolean,
-    forcedPackageType: Option[PackageType],
+    forcedPackageTypeOpt: Option[PackageType],
     build: Build.Successful,
     extraArgs: Seq[String],
     expectedModifyEpochSecondOpt: Option[Long]
   ): Either[BuildException, Option[Long]] = either {
 
-    val packageType = forcedPackageType.getOrElse {
-      // FIXME We'll probably need more refined rules if we start to support extra Scala.js or Scala Native specific types
-      if (build.options.notForBloopOptions.packageOptions.isDockerEnabled)
-        PackageType.Docker
-      else if (build.options.platform.value == Platform.JS)
-        PackageType.Js
-      else if (build.options.platform.value == Platform.Native)
-        PackageType.Native
-      else
-        build.options.notForBloopOptions.packageOptions.packageTypeOpt
-          .getOrElse(PackageType.Bootstrap)
+    val packageType: PackageType = {
+      val basePackageTypeOpt = build.options.notForBloopOptions.packageOptions.packageTypeOpt
+      lazy val validPackageScalaJS =
+        Seq(PackageType.LibraryJar, PackageType.SourceJar, PackageType.DocJar)
+      lazy val validPackageScalaNative =
+        Seq(PackageType.LibraryJar, PackageType.SourceJar, PackageType.DocJar)
+
+      (forcedPackageTypeOpt -> build.options.platform.value) match {
+        case (Some(forcedPackageType), _) => forcedPackageType
+        case (_, _) if build.options.notForBloopOptions.packageOptions.isDockerEnabled =>
+          for (basePackageType <- basePackageTypeOpt)
+            throw new MalformedCliInputError(
+              s"Unsuported package type: $basePackageType for Docker."
+            )
+          PackageType.Docker
+        case (_, Platform.JS) =>
+          val validatedPackageType =
+            for (basePackageType <- basePackageTypeOpt)
+              yield
+                if (validPackageScalaJS.contains(basePackageType)) basePackageType
+                else throw new MalformedCliInputError(
```

Comment:
As much as possible, we shouldn't throw here, but rather try to make the exception be returned via the left side of the either. Thrown exceptions aren't reported as nicely as the ones we propagate via eithers.